CSP-2215 Implements Add provider - offers apprenticeship question

afzal-hussain259 opened the pull request.

afzal-hussain259 committed:
CSP-2215 Implement offer apprenticeships page

CSP-2215 Unit tests added

CSP-2215 Updated button

CSP-2215 Corrected spelling

CSP-2215 Addressed comments

DevenKShah commented on src/SFA.DAS.Admin.Roatp.Web/Controllers/AddProvider/SelectOfferApprenticeshipsController.cs:
```diff
@@ -0,0 +1,69 @@
+            foreach (var error in result.Errors)
+            {
+                ModelState.AddModelError(error.PropertyName, error.ErrorMessage);
+            }
```
This is a repetitive code, it should be converted in to an extension

DevenKShah commented on src/SFA.DAS.Admin.Roatp.Web/Controllers/AddProvider/SelectProviderTypeController.cs:
```diff
@@ -54,7 +54,19 @@ public IActionResult Index(SelectProviderTypeSubmitModel submitModel)
-        return RedirectToRoute(RouteNames.SelectProviderType);
+        if (sessionModel.OfferApprenticeships != null)
+        {
+            sessionModel.OfferApprenticeships = null;
```
As you move along in the journey, more fields will be required to clear. If you implement a method in the session model to clear fields and invoke it here, then this code will work with out any further change.

DevenKShah commented on src/SFA.DAS.Admin.Roatp.Web/Controllers/ApprenticeshipsUpdateController.cs:
Remember to add notes on the story when you make impactful changes on existing bits so QAs can make sure they test this as well.

afzal-hussain259 committed:
CSP-2216 Implements offer apprenticeship units page

CSP-2216 fixed small routing

CSP-2216 Addressed further comments

CSP-2216 Updated unit tests

DevenKShah commented on src/SFA.DAS.Admin.Roatp.Web/Extensions/ModelStateExtensions.cs:
```diff
@@ -0,0 +1,15 @@
+
+public static class ModelStateExtensions
+{
+    public static void AddValidationErrors(this ModelStateDictionary modelState, IEnumerable<ValidationFailure> errors)
```
Excellent

DevenKShah commented on src/SFA.DAS.Admin.Roatp.Web/Controllers/AddProvider/SelectProviderController.cs:
```diff
@@ -56,13 +55,23 @@ public async Task<IActionResult> Index(SelectProviderSubmitModel submitModel, Ca
             return View(viewModel);
         }
 
+        if (organisationApiResponse.StatusCode != System.Net.HttpStatusCode.OK && organisationApiResponse.StatusCode != System.Net.HttpStatusCode.NotFound)
```
`organisationApiResponse.StatusCode != System.Net.HttpStatusCode.OK` is redundant check

DevenKShah commented on src/SFA.DAS.Admin.Roatp.Web/Controllers/AddProvider/SelectProviderTypeController.cs:
It is possible that session has timed out. The session object will return null and user will see an error. A session check has to be carried out first thing in post just like you have done in get. Also the session check is becoming repetitive and has no business benefit, so this is a candidate for cross cutting concern and it can be moved out of controller action using a filter.

DevenKShah commented on src/SFA.DAS.Admin.Roatp.Web/Models/Session/AddProviderSessionModel.cs:
```diff
@@ -7,4 +7,11 @@ public class AddProviderSessionModel
     public int? ProviderTypeId { get; set; }
+    public bool? OfferApprenticeships { get; set; }
+    public bool? OfferApprenticeshipUnits { get; set; }
+    public void ClearSessionProperty(string propertyName)
```
This is clever. However, you will require to call these multiple times as more properties get added. I would just create a simple method as follows
```
public void ResetModel() 
{
  OfferApprenticeships = null;
  //add others as they come along
}
```

DevenKShah commented on src/SFA.DAS.Admin.Roatp.Web/Controllers/AddProvider/SelectOfferApprenticeshipsController.cs:
```diff
@@ -0,0 +1,79 @@
+            return RedirectToRoute(RouteNames.SelectOfferApprenticeships);
+        }
+
+        return RedirectToRoute(RouteNames.SelectOfferApprenticeshipUnits);
```
Irrespective of the answer the navigation remains same isn't it?

DevenKShah commented on src/SFA.DAS.Admin.Roatp.Web/Controllers/AddProvider/SelectOfferApprenticeshipsController.cs:
```diff
@@ -0,0 +1,79 @@
+        {
+            sessionModel.ClearSessionProperty(nameof(sessionModel.OfferApprenticeshipUnits));
+
+            _sessionService.Set(SessionKeys.AddProvider, sessionModel);
```
I dont get this, why are we clearing units question? and why are we setting session model multiple times, you can make changes to the session model as you go, but set it only ones at the end.

afzal-hussain259 committed:
CSP-2369 Update coursetype model

afzal-hussain259 committed:
CSP-2215 Renamed properties

DevenKShah commented on src/SFA.DAS.Admin.Roatp.Web/AppStart/NoCacheMiddlewareExtensions.cs:
```diff
@@ -0,0 +1,14 @@
+[ExcludeFromCodeCoverage]
+public static class NoCacheMiddlewareExtensions
+{
+    public static IApplicationBuilder UseNoBrowserCache(this IApplicationBuilder builder)
```
There is no need for this extension, just add UseMiddleware directly in start up

DevenKShah commented on src/SFA.DAS.Admin.Roatp.Web/Program.cs:
```diff
@@ -78,6 +79,7 @@
     .UseStaticFiles()
     .UseCookiePolicy()
     .UseRouting()
+    .UseNoBrowserCache()
```
This should be placed  before UseRouting

DevenKShah commented on src/SFA.DAS.Admin.Roatp.Web/Models/OfferApprenticeshipUnitsViewModel.cs:
```diff
@@ -0,0 +1,13 @@
+
+public class OfferApprenticeshipUnitsSubmitModel
+{
+    public bool? ApprenticeshipUnitsSelectionId { get; set; }
```
Since this is a boolean, the variable name should be a question ie. `IsApprenticeshipUnitsOffered`

DevenKShah commented on src/SFA.DAS.Admin.Roatp.Web/Models/OfferApprenticeshipsViewModel.cs:
```diff
@@ -2,8 +2,12 @@
+
+public class OfferApprenticeshipsSubmitModel
+{
     public bool? ApprenticeshipsSelectionChoice { get; set; }
```
Since this is a boolean, the variable name should be a question ie. IsApprenticeshipsOffered

DevenKShah commented on src/SFA.DAS.Admin.Roatp.Web/Models/Session/AddProviderSessionModel.cs:
```diff
@@ -7,4 +7,11 @@ public class AddProviderSessionModel
     public string? CharityNumber { get; set; }
     public string? CompanyNumber { get; set; }
     public int? ProviderTypeId { get; set; }
+    public bool? OfferApprenticeships { get; set; }
```
Use Offers instead of Offer so that it becomes a question
